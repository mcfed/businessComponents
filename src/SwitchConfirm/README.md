## SwitchConfirm component

useage

```jsx
import {SwitchConfirm} from '@mcf/components';
class App extends React.Component {
  handleChangeSwitch(checked: boolean, action: () => void) {
    Modal.confirm({
      title: checked ? '111' : '222',
      onOk: () => {
        action();
      }
    });
  }
  render() {
    return (
      <SwitchConfirm
        onSwitchChange={this.handleChangeSwitch.bind(this)}></SwitchConfirm>
    );
  }
}
```

## Options

> onSwitchChange 作为改变回调函数 代替 onChange 接收到 checked （状态） 和 后续操作 action 若希望 switch 状态改变 则需要调用 action

| 参数           | 说明           | 类型                     | 默认值 |
| -------------- | -------------- | ------------------------ | ------ |
| onSwitchChange | 必填，改变函数 | function(checked,action) | -      |
