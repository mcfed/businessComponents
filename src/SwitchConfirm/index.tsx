import * as React from 'react';
// import Switch, {SwitchProps} from 'antd/es/switch';
// import Modal, {ModalFuncProps} from 'antd/es/modal';
import {SwitchProps} from 'antd/es/switch';
import {ModalFuncProps} from 'antd/es/modal';
import {Switch, Modal} from 'antd';

interface SwitchConfirmProps extends SwitchProps {
  /**
   * 自定义非选中时的值，默认为false
   */
  uncheckedOption?: any;
  /**
   * 自定义选中时的值，默认为true
   */
  checkedOption?: any;
  /**
   * 当前状态
   */
  currentOption?: any;
  /**
   * 弹出框确认按钮的回调
   */
  onConfirm: (currentOption: any, action?: () => void) => void;
  /**
   * 暂无用意
   */
  confirm?: boolean;
  /**
   * 弹出框的props属性
   */
  modalConfirmProps?: ModalFuncProps;
}
interface SwitchConfirmState {
  checked: boolean;
}

export default class SwitchConfirm extends React.Component<
  SwitchConfirmProps,
  SwitchConfirmState
> {
  static defaultProps = {
    prefixCls: 'ant-switch'
  };
  constructor(props: SwitchConfirmProps) {
    super(props);
    this.state = {
      checked: this.translateChecked(props, props.checked)
    };
  }
  // componentDidMount() {
  //   this.setState({
  //     checked: this.translateChecked(this.props, this.props.checked)
  //   });
  // }
  componentWillReceiveProps(nextProps: SwitchConfirmProps) {
    if (
      this.props.checked !== nextProps.checked ||
      this.props.currentOption !== nextProps.currentOption
    ) {
      this.setState({
        checked: this.translateChecked(nextProps, nextProps.checked)
      });
    }
  }
  translateChecked(currentProps: SwitchConfirmProps, checkedProp?: boolean) {
    const {currentOption, checkedOption} = currentProps;
    if (currentOption !== undefined && checkedOption !== undefined) {
      return currentOption == checkedOption;
    }
    return checkedProp !== undefined ? checkedProp : false;
  }
  translateCustomChecked(checked: boolean) {
    const {uncheckedOption, checkedOption} = this.props;
    if (uncheckedOption !== undefined && checkedOption !== undefined) {
      return checked ? checkedOption : uncheckedOption;
    }
    return checked;
  }
  handleChange(checked: boolean) {
    const _this = this;
    const {modalConfirmProps, onConfirm} = this.props;
    const currentOption = this.translateCustomChecked(checked);
    Modal.confirm({
      ...modalConfirmProps,
      onOk: () => {
        onConfirm(currentOption, () =>
          _this.setState({
            checked: checked
          })
        );
      }
    });
  }

  render() {
    const {
      confirm,
      modalConfirmProps,
      onChange,
      onConfirm,
      currentOption,
      checkedOption,
      uncheckedOption,
      ...otherProps
    } = this.props;
    const {checked} = this.state;
    return (
      <Switch
        {...otherProps}
        checked={checked}
        onChange={this.handleChange.bind(this)}
      />
    );
  }
}
