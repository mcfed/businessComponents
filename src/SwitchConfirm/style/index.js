import 'antd/es/switch/style';
import 'antd/es/modal/style';
