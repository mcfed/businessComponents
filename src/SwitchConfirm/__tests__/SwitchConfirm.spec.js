import React from 'react';
import {shallow, mount} from 'enzyme';

import SwitchConfirm from '../index';
import {Modal} from 'antd';

Modal.confirm = jest.fn();

const setup = cprops => {
  const defaultProps = {
    confirm: jest.fn()
  };
  const props = Object.assign({}, defaultProps, cprops);
  const wrapper = shallow(<SwitchConfirm {...props} />);
  return {
    wrapper,
    props
  };
};

describe('快照测试', () => {
  it('switchconfirm  ', () => {
    const {wrapper} = setup();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('switchconfirm method test', () => {
  it('translateChecked test', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();

    expect(instance.translateChecked({}, true)).toEqual(true);
    expect(instance.translateChecked({}, false)).toEqual(false);
    expect(instance.translateChecked({}, undefined)).toEqual(false);

    expect(
      instance.translateChecked(
        {
          currentOption: 'on',
          checkedOption: 'on'
        },
        undefined
      )
    ).toEqual(true);
    expect(
      instance.translateChecked(
        {
          currentOption: 'off',
          checkedOption: 'on'
        },
        undefined
      )
    ).toEqual(false);
  });

  it('translateCustomChecked test 1', () => {
    const {wrapper, props} = setup({
      uncheckedOption: 'off',
      checkedOption: 'on'
    });
    const instance = wrapper.instance();

    expect(instance.translateCustomChecked(true)).toEqual(props.checkedOption);
    expect(instance.translateCustomChecked(false)).toEqual(
      props.uncheckedOption
    );
  });

  it('translateCustomChecked test 2', () => {
    const {wrapper, props} = setup();
    const instance = wrapper.instance();

    expect(instance.translateCustomChecked(true)).toEqual(true);
    expect(instance.translateCustomChecked(false)).toEqual(false);
  });

  // it('componentDidMount test 1', () => {
  //   const {wrapper, props} = setup();
  //   const instance = wrapper.instance();

  //   instance.componentDidMount();
  //   expect(wrapper.state('checked')).toEqual(false);
  // });
  // it('componentDidMount test 2', () => {
  //   const {wrapper, props} = setup({
  //     checked: true
  //   });
  //   const instance = wrapper.instance();

  //   instance.componentDidMount();
  //   expect(wrapper.state('checked')).toEqual(true);
  // });

  it('componentWillReceiveProps test 1', () => {
    const {wrapper, props} = setup();
    const instance = wrapper.instance();

    instance.componentWillReceiveProps({
      checked: true,
      currentOption: 'on'
    });
    expect(wrapper.state('checked')).toEqual(true);
  });
  it('componentWillReceiveProps test 2', () => {
    const {wrapper, props} = setup();
    const instance = wrapper.instance();

    instance.componentWillReceiveProps(props);
    expect(wrapper.state('checked')).toEqual(false);
  });

  it('handleChange test 1', () => {
    const {wrapper, props} = setup();
    const instance = wrapper.instance();

    expect(instance.handleChange(true)).toEqual(undefined);
    expect(Modal.confirm).toHaveBeenCalled();
  });
});
