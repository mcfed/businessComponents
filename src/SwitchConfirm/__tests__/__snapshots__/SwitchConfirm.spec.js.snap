// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 switchconfirm   1`] = `
<Switch
  checked={false}
  onChange={[Function]}
  prefixCls="ant-switch"
/>
`;
